### PR TITLE
SLA Report

### PR DESCRIPTION
**Pull request by nikhilmn**
First commit.

### CODE REVIEW

File: sla-miss-report/sla_miss_report.py [hunk 1]
```diff
@@ -0,0 +1,902 @@
...
+    Text,
+    create_engine,
+    func,
+    not_,
+    or_,
+)
+from sqlalchemy.orm import backref, joinedload, relationship
+from sqlalchemy.orm.query import Query
+from sqlalchemy.orm.session import Session
+from yaml import serialize
```
Comment (prakshalj0512) [inline sla-miss-report/sla_miss_report.py:33]: remove all unused packages

File: sla-miss-report/sla_miss_report.py [hunk 2]
```diff
@@ -0,0 +1,902 @@
...
+
+session = settings.Session()
+
+slamiss = session.query(
```
Comment (prakshalj0512) [inline sla-miss-report/sla_miss_report.py:37]: remove all references to SlaMiss table

File: sla-miss-report/sla_miss_report.py [hunk 3]
```diff
@@ -0,0 +1,902 @@
...
+taskinstance_df["run_date_hour"] = pd.to_datetime(taskinstance_df["start_date"]).dt.hour
+taskinstance_df["task_queue_time"] = (
+    taskinstance_df["start_date"] - taskinstance_df["queued_dttm"]
+).dt.total_seconds()
+taskinstance_df = taskinstance_df[taskinstance_df["task_queue_time"] > 0]
+
+dagrun = session.query(DagRun.dag_id, DagRun.run_id, DagRun.data_interval_end).all()
+dagrun_df = pd.DataFrame(dagrun)
+dagrun_df.rename(columns={"data_interval_end": "actual_start_time"}, inplace=True)
+# print(dagrun_df[:10])
```
Comment (prakshalj0512) [inline sla-miss-report/sla_miss_report.py:66]: remove all unnecessary code comments

File: sla-miss-report/sla_miss_report.py [hunk 4]
```diff
@@ -0,0 +1,902 @@
...
+# print(sla_highest_sla_miss_hour)
+# print(sla_highest_tasks_hour)
+# print(sla_longest_queue_time_hour)
+
+
+# print(sla_miss_percent_past_day_hourly)
+
+# SLA_Detailed
+
+time_diff_7days = timedelta(days=7, hours=0, minutes=0)
```
Comment (prakshalj0512) [inline sla-miss-report/sla_miss_report.py:459]: reuse the vars from earlier

```
time_diff_one_week = timedelta(days=7, hours=0, minutes=0)
time_diff_one_day = timedelta(days=7, hours=0, minutes=0)
time_diff_7days = timedelta(days=7, hours=0, minutes=0)
```

File: sla-miss-report/sla_miss_report.py [hunk 5]
```diff
@@ -0,0 +1,902 @@
...
+taskinstance_df = taskinstance_df[taskinstance_df["task_queue_time"] > 0]
+
+dagrun = session.query(DagRun.dag_id, DagRun.run_id, DagRun.data_interval_end).all()
+dagrun_df = pd.DataFrame(dagrun)
+dagrun_df.rename(columns={"data_interval_end": "actual_start_time"}, inplace=True)
+# print(dagrun_df[:10])
+# dagrun.rename(columns={'gdp':'log(gdp)'}, inplace=True)
+# taskinstance_df.info()
+
+serializeddag = session.query(SerializedDagModel._data).all()
```
Comment (prakshalj0512) [inline sla-miss-report/sla_miss_report.py:70]: change the `_data` to `data` and see if it works. _data failed to work in v2.2.0.

File: sla-miss-report/sla_miss_report.py [hunk 6]
```diff
@@ -0,0 +1,902 @@
...
+serializeddag_dfjson = pd.DataFrame(serializeddag)
+df2 = serializeddag_dfjson["_data"].apply(json.dumps)
```
Comment (prakshalj0512) [inline sla-miss-report/sla_miss_report.py:72]: rename df to something meaningful

File: sla-miss-report/sla_miss_report.py [hunk 7]
```diff
@@ -0,0 +1,902 @@
...
+df3 = df2.apply(json.loads)
+df4 = pd.DataFrame(df3.values.tolist())["dag"]
```
Comment (prakshalj0512) [inline sla-miss-report/sla_miss_report.py:74]: try to see if you can consolidate all these transformations (line 71-74) into one line

**Review by prakshalj0512 — CHANGES_REQUESTED** [review]
great job on the code! here are a few general suggestions in addition to inline comments:

- [ ] rename PR title to be more relevant
- [ ] remove any & every unused line of code
- [ ] create functions for each different metric and return df & recommendation for use later
- [ ] print out the returned output of each func in the script so it shows up in airflow task logs
- [ ] add docstrings to those functions 
- [ ] Create global variables for things like email list. Check out existing DAGs for reference. 
- [ ] create a func for concatenating multiple strings/df cols
- [ ] parameterize the days (1,3,7)
- [ ] add checks for no data  
- [ ] add if/else clause for _data in serializedDAG

File: sla-miss-report/sla_miss_report.py [hunk 8]
```diff
@@ -0,0 +1,902 @@
...
+    description="A simple email ",
+    schedule_interval=None,
+    start_date=datetime(2021, 1, 1),
+    catchup=False,
+) as dag:
+
+    send_email_notification = EmailOperator(
+        task_id="send_test_email",
+        to=["nikhil.manjunatha@clairvoyantsoft.com"],
+        subject="Airflow SLA Metrics Report - 08/04/2022",
```
Comment (prakshalj0512) [inline sla-miss-report/sla_miss_report.py:837]: generate date dynamically

File: sla-miss-report/sla_miss_report.py [hunk 9]
```diff
@@ -0,0 +1,902 @@
...
+        "sla_miss_percent(missed_tasks/total_tasks)": "SLA Miss % (missed_tasks/total_tasks)",
+    },
+    inplace=True,
+)
+
+# print(slamiss_pct_last7days)
+
+# Hourly Trend
+
+time_diff_one_day = timedelta(days=7, hours=0, minutes=0)
```
Comment (prakshalj0512) [inline sla-miss-report/sla_miss_report.py:252]: revert back to 1 day

**Comment by prakshalj0512** [comment]
- [ ] Remove methods that cant be reused
- [ ] Rename any DF vars to have either _dag_, _daily_, _hourly
- [ ] Remove hardcoded 1,3,7 days
- [ ] Loop thru the days list wherever it’s being used; try
- [ ] Rename df2-4
- [ ] Try to see if ‘yesterday’ can be used instead of today min time
- [ ] Func for obs5_sladpercent_weekprior, obs5_sladpercent_weekprior
- [ ] Reword the descriptions to be more direct & authorotative
- [ ] Test returning empty DF to jinja templating; adding a check for DF len in templating
- [ ] Format the code before pushing
- [ ] Remove all unused code (SLAMiss table)
- [ ] Create vars for list of days
```
short_time_frame = 1
medium_time_frame = 3
long_time_frame = 7

list_days = [short_time_frame, medium_time_frame, long_time_frame]
```

**Review by prakshalj0512 — CHANGES_REQUESTED** [review]
Please review the comments. Reference for common issues:
https://clairvoyant.atlassian.net/wiki/spaces/INSIGHT/pages/2749071361/Git+Peer+Reviews

File: sla-miss-report/sla_miss_report.py [hunk 10]
```diff
@@ -0,0 +1,877 @@
...
+
+        if "_data" in dir(SerializedDagModel):
+            serializeddag = session.query(SerializedDagModel._data).all()
+            data_col = "_data"
+        else:
+            serializeddag = session.query(SerializedDagModel.data).all()
+            data_col = "data"
+
+        serializeddag_df = pd.DataFrame(serializeddag)
+        serializeddag_json = serializeddag_df["_data"].apply(json.dumps).apply(json.loads)
```
[the remark this answers is not in the document]
Comment (nikhilmn) [reply sla-miss-report/sla_miss_report.py]: resolved

**Review by prakshalj0512 — APPROVED** [review]

**Pull request merged**